File: test/ModularPipelines.UnitTests/Attributes/ModuleDependencyRegistryTests.cs
using ModularPipelines.Context;
using ModularPipelines.Engine.Dependencies;
using ModularPipelines.Modules;

namespace ModularPipelines.UnitTests.Attributes;

public class ModuleDependencyRegistryTests
{
    private class ModuleA : Module<string>
    {
        public override Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
            => Task.FromResult<string?>("A");
    }

    private class ModuleB : Module<string>
    {
        public override Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
            => Task.FromResult<string?>("B");
    }

    private class ModuleC : Module<string>
    {
        public override Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
            => Task.FromResult<string?>("C");
    }

    [Test]
    public async Task AddDynamicDependency_AddsDependency()
    {
        var registry = new ModuleDependencyRegistry();

        registry.AddDynamicDependency(typeof(ModuleA), typeof(ModuleB));

        var dependencies = registry.GetDynamicDependencies(typeof(ModuleA));
        await Assert.That(dependencies).Contains(typeof(ModuleB));
    }

    [Test]
    public async Task AddDynamicDependency_MultipleDependencies_AddsAll()
    {
        var registry = new ModuleDependencyRegistry();

        registry.AddDynamicDependency(typeof(ModuleA), typeof(ModuleB));
        registry.AddDynamicDependency(typeof(ModuleA), typeof(ModuleC));

        var dependencies = registry.GetDynamicDependencies(typeof(ModuleA));
        await Assert.That(dependencies.Count()).IsEqualTo(2);
        await Assert.That(dependencies).Contains(typeof(ModuleB));
        await Assert.That(dependencies).Contains(typeof(ModuleC));
    }

    [Test]
    public async Task RemoveDependency_RemovesDependency()
    {
        var registry = new ModuleDependencyRegistry();

        registry.AddDynamicDependency(typeof(ModuleA), typeof(ModuleB));
        registry.RemoveDependency(typeof(ModuleA), typeof(ModuleB));

        var dependencies = registry.GetDynamicDependencies(typeof(ModuleA));
        await Assert.That(dependencies).IsEmpty();
    }

    [Test]
    public async Task GetDynamicDependencies_NoDependencies_ReturnsEmpty()
    {
        var registry = new ModuleDependencyRegistry();

        var dependencies = registry.GetDynamicDependencies(typeof(ModuleA));
        await Assert.That(dependencies).IsEmpty();
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Attributes/ModuleDependencyRegistryTests.fs
namespace ModularPipelines.UnitTests.FSharp.Attributes

open System.Threading
open System.Threading.Tasks
open ModularPipelines.Context
open ModularPipelines.Engine.Dependencies
open ModularPipelines.Modules
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

module ModuleDependencyRegistryTests =
    type ModuleA() =
        inherit Module<string>()
        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            Task.FromResult<string>("A")

    type ModuleB() =
        inherit Module<string>()
        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            Task.FromResult<string>("B")

    type ModuleC() =
        inherit Module<string>()
        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            Task.FromResult<string>("C")

type ModuleDependencyRegistryTests() =
    [<Test>]
    member _.AddDynamicDependency_AddsDependency() = async {
        let registry = ModuleDependencyRegistry()
        registry.AddDynamicDependency(typeof<ModuleDependencyRegistryTests.ModuleA>, typeof<ModuleDependencyRegistryTests.ModuleB>)
        let dependencies = registry.GetDynamicDependencies(typeof<ModuleDependencyRegistryTests.ModuleA>)
        do! check(Assert.That(dependencies |> Seq.contains typeof<ModuleDependencyRegistryTests.ModuleB>).IsTrue())
    }

    [<Test>]
    member _.AddDynamicDependency_MultipleDependencies_AddsAll() = async {
        let registry = ModuleDependencyRegistry()
        registry.AddDynamicDependency(typeof<ModuleDependencyRegistryTests.ModuleA>, typeof<ModuleDependencyRegistryTests.ModuleB>)
        registry.AddDynamicDependency(typeof<ModuleDependencyRegistryTests.ModuleA>, typeof<ModuleDependencyRegistryTests.ModuleC>)
        let dependencies = registry.GetDynamicDependencies(typeof<ModuleDependencyRegistryTests.ModuleA>)
        do! check(Assert.That((dependencies |> Seq.length) = 2).IsTrue())
        do! check(Assert.That(dependencies |> Seq.contains typeof<ModuleDependencyRegistryTests.ModuleB>).IsTrue())
        do! check(Assert.That(dependencies |> Seq.contains typeof<ModuleDependencyRegistryTests.ModuleC>).IsTrue())
    }

    [<Test>]
    member _.RemoveDependency_RemovesDependency() = async {
        let registry = ModuleDependencyRegistry()
        registry.AddDynamicDependency(typeof<ModuleDependencyRegistryTests.ModuleA>, typeof<ModuleDependencyRegistryTests.ModuleB>)
        registry.RemoveDependency(typeof<ModuleDependencyRegistryTests.ModuleA>, typeof<ModuleDependencyRegistryTests.ModuleB>)
        let dependencies = registry.GetDynamicDependencies(typeof<ModuleDependencyRegistryTests.ModuleA>)
        do! check(Assert.That(Seq.isEmpty dependencies).IsTrue())
    }

    [<Test>]
    member _.GetDynamicDependencies_NoDependencies_ReturnsEmpty() = async {
        let registry = ModuleDependencyRegistry()
        let dependencies = registry.GetDynamicDependencies(typeof<ModuleDependencyRegistryTests.ModuleA>)
        do! check(Assert.That(Seq.isEmpty dependencies).IsTrue())
    }
